### ISSUE TITLE
Merging a dynamic into the view model

### ISSUE THREAD

**Issue by MarcusChamberlain** [ServiceComposer/ServiceComposer.AspNetCore#32]
Hi there

I've been testing Service Composer and ran into a small issue.

The 'viewmodel' property used in the query handlers is not an ExpandoObject, it is of type  ServiceComposer.DynamicViewModel. This hides the 'properties' dictionary in a private field in an internal class. The issue is that I have a dynamic already returned from a call to a nosql database. I can't assign this directly to the viewmodel (ie, vm = noSqlResponse), so I would like to iterate over the keys in the dictionary and add the new properties to the vm. Usually with an ExpandoObject I would do something like:

```
        dynamic thing1 = new ExpandoObject();
        dynamic thing2 = new ExpandoObject();
        thing1.Bar = "Foo";
        thing2.Foo = "Bar";
        Merge(thing1, thing2);
        //result: thing2 now contains .Bar and .Foo

        private void Merge(IDictionary<string, object> from, IDictionary<string, object> to)
        {
            foreach (var key in from.Keys)
                to.Add(new KeyValuePair<string, object>(key, from[key]));
        }
```

But since I can't access the dictionary behind the DynamicViewModel I can't add new records. Likewise, if I try to set a property directly at runtime (ie. not using the 'vm.NewProperty = 'Foo' syntax)

`((IDictionary<String, Object>) vm)[propKey] = propValue;`

I get an exception that type DynamicViewModel can't be cast to type IDictionary, which isn't a surprise.

So it seems that I'm pretty well locked out of DynamicViewModel. As a workaround I can assign a temporary value like "vm.response = new ExpandoObject()" and then assign my noSqlResponse. That works, but it's a bit odd to nest all returned data under a single property - I'd much rather just return the data on the root. 

Any suggestions on how I might access the properties dictionary in DynamicViewModel to add new properties at runtime to the view model root?

Thanks.

**Comment by mauroservienti** [comment]
Hi @MarcusChamberlain, thanks for raising  this. Never thought about that scenario, indeed sounds interesting and doable. I see to options:

- `DynamicViewModel` has a `Merge` method 
- `DynamicViewModel` implements `IDictionary<>`

Thoughts?

**Comment by MarcusChamberlain** [comment]
Thanks for taking a look into this. Being able to call 'vm.Merge(myDynamic)' on the view model would be very nice.

**Comment by mauroservienti** [comment]
@MarcusChamberlain I raised https://github.com/ServiceComposer/ServiceComposer.AspNetCore/pull/33 to address this. The only thing I'm not sure about is the `Merge` behavior, right now I took the simplest approach that is to overwrite existing properties in `DynamicViewModel`. Obviously this can lead to some strange scenarios like:

```
dynamic d = new DynamicViewModel(....);
d.SomeProperty = 10;

dynamic source = new ExpandoObject();
source.SomeProperty = "some value";

d.Merge(source);
```

It'll work but now the `DynamicViewModel` instance `SomeProperty` type changed from `int` to `string`.

I could check that in case of overwrite the type is the same.

Thoughts?

**Comment by MarcusChamberlain** [comment]
Hi Mauro

The merge solution looks great. I did notice one issue though.

At the moment your unit tests (in the 'When_merging_to_DynamicViewModel' class) are able to run because you have a compiler instruction to make internal classes visible to the external assembly 'ServiceComposer.AspNetCore.Tests'. This is tricking your tests into passing - if you attempt to use vm.Merge() in an project without that compiler instruction, you'll get an error that DynamicViewModel does not contain a method called Merge(). 

This occurs because the call to Merge() goes through the 'TryInvokeMember' method. This has been set to only allow calls to 'RaiseEvents'.

The fix is to add the following to 'TryInvokeMember':

```    
            if (binder.Name == "Merge")
            {
                this.Merge(args[0] as IDictionary<string, object>);
                return true;
            }
```

**Comment by mauroservienti** [comment]
Great catch @MarcusChamberlain. I'll get it fixed. If I end up using the same approach as for `RaiseEvent` I can also drop the `IDynamicViewModel` interface entirely.

**Comment by mauroservienti** [comment]
@MarcusChamberlain I just pushed a fix to #33

**Comment by mauroservienti** [comment]
Merged the PR and deployed `0.0.11` to Nuget. Thanks @MarcusChamberlain.
